katsdpimager/primary_beam.py

```diff
@@ -208,59 +62,127 @@ class TrivialBeamModel(BeamModel):
 
+    @property
+    def is_circular(self) -> bool:
+        return True
 
-class MeerkatBeamModelSet1(BeamModelSet):
-    """A very simple model of the MeerKAT dish beam."""
+    @property
+    def is_unpolarized(self) -> bool:
+        return True
 
-    BANDS = {'L', 'UHF'}
+    @property
+    def antenna(self) -> None:
+        return None
 
-    def __init__(self, band: str) -> None:
-        if band not in self.BANDS:
-            raise ValueError(f'band ({band}) must be one of {self.BANDS}')
-        filename = pkg_resources.resource_filename(
-            'katsdpimager',
-            f'models/beams/meerkat/v1/beam_{band}.h5')
-        self._model = TrivialBeamModel(filename)
+    @property
+    def receiver(self) -> None:
+        return None
 
     @property
-    def parameters(self) -> Sequence[Parameter]:
-        return []
+    def band(self) -> str:
+        return self._band
+
+    def sample(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,
+               frame: Union[AltAzFrame, RADecFrame],
+               output_type: OutputType, *,
+               out: Optional[np.ndarray] = None) -> np.ndarray:
+        l_ = np.asarray(l)
+        m_ = np.asarray(m)
+        if output_type != OutputType.UNPOLARIZED_POWER:
+            raise NotImplementedError('Only UNPOLARIZED_POWER is currently implemented')
+        in_shape = np.broadcast_shapes(l_.shape, m_.shape, frame.shape)
+        out_shape = frequency.shape + in_shape
+        if out is not None:
+            if out.shape != out_shape:
+                raise ValueError(f'out must have shape {out_shape}, not {out.shape}')
+            if out.dtype != np.float32:
+                raise TypeError(f'out must have dtype float32, not {out.dtype}')
+            if not out.flags.c_contiguous:
+                raise ValueError('out must be C contiguous')
+        else:
+            out = np.empty(out_shape, np.float32)
+        frequency_Hz = frequency.to_value(u.Hz).astype(np.float32, copy=False, casting='same_kind')
+        samples = self._interp_samples(frequency_Hz)
+        # Create view with l/m axis flattened to 1D for benefit of numba
+        l_view = np.broadcast_to(l_, in_shape).ravel()
+        m_view = np.broadcast_to(m_, in_shape).ravel()
+        out_view = out.view()
+        out_view.shape = frequency.shape + l_view.shape
+        _sample_impl(l_view, m_view, samples, self._step, out_view)
+        return out
 
-    def sample(self, **kwargs) -> BeamModel:
-        return self._model
+    def sample_grid(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,
+                    frame: Union[AltAzFrame, RADecFrame],
+                    output_type: OutputType, *,
+                    out: Optional[np.ndarray] = None) -> np.ndarray:
+        l_ = np.asarray(l)
+        m_ = np.asarray(m)
+        return self.sample(
+            l_[np.newaxis, :], m_[:, np.newaxis], frequency,
+            frame, output_type, out=out
+        )
+
+
+BANDS = {'L', 'UHF'}
```

Comment:
might we ever support 'S' band TrivialPrimaryBeams

Reply:
Potentially - it depends when we get demand for PB-corrected S-band images versus when katsdpimager supports non-trivial primary beams.